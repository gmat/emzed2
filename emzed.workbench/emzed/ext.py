import pkg_resources

for ep in pkg_resources.iter_entry_points("emzed_plugin", name="package"):
    print ep
    package = ep.load()
    exec "%s=package" % package.__name__
    del package
    del ep
del pkg_resources

