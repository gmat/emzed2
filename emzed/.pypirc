[distutils]
index-servers = 
  local

[local]
repository: http://127.0.0.1:8080
username:uschmitt
password:pillepalle

