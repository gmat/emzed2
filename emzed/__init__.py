# encoding:latin-1

from version import version as __version__


