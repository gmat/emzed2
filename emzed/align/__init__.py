from rt_align import rtAlign
from mz_align import mzAlign
