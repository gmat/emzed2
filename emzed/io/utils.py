

def _prepare_path(path, extensions):

    import sys
    if isinstance(path, unicode):
        path = path.encode(sys.getfilesystemencoding())
    elif path is None:
        from .. import gui
        path = gui.askForSingleFile(extensions=extensions)
    return path
