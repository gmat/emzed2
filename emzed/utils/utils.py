import emzed.core.data_types

def toTable(colName, iterable,  format_="", type_=None, title="", meta=None):
    return emzed.core.data_types.Table.toTable(colName, iterable, format_, type_, title, meta)

toTable.__doc__ = emzed.core.data_types.Table.toTable.__doc__

def formula(mf):
    """
    Creates formula object which allows addition and subtraction:

    .. pycon::

       import ms   !onlyoutput
       mf1 = ms.formula("H2O")
       mf2 = ms.formula("NaOH")
       mf3 = mf1 + mf2
       print str(mf3)
       print str(mf3 - mf1)

    Mass calculation is supported too:

    .. pycon::

       print mf1.mass()

    If you need some internal representation, you can get a dictionary.
    Keys are pairs of *(symbol, massnumber)*, where *massnumber = None*
    refers to the lowest massnumber. Values of the dictionary are counts:

    .. pycon::

       print mf1.asDict()
       mixed = ms.formula("[13]C2[14]C3")
       print mixed.asDict()


    """
    from emzed.core.chemistry import MolecularFormula
    return MolecularFormula(mf)

def addmf(formula0, *formulas):
    """
    Combines molecular formulas by addition and subtraction:

    .. pycon::

       import ms !onlyoutput
       print ms.addmf("H2O", "COOH")
       print ms.addmf("H2O", "COOH", "NaCl")

    A leading minus sign subtracts the formula following this sign:

    .. pycon::

       print ms.addmf("H2O2", "-H2O")
       print ms.addmf("H2O", "COOH", "NaCl", "-H2O2")
       print ms.addmf("(CH2)7COOH", "-C7")

    """

    mf0 = formula(formula0)
    for f in formulas:
        if f.startswith("-"):
            mf0 -= formula(f[1:])
        else:
            mf0 += formula(f)
    return str(mf0)


###


def openInBrowser(urlPath):
    """
    opens *urlPath* in browser, eg:

    .. pycon::
        ms.openInBrowser("http://emzed.biol.ethz.ch") !noexec

    """
    from PyQt4.QtGui import QDesktopServices
    from PyQt4.QtCore import QUrl
    import os.path

    url = QUrl(urlPath)
    scheme = url.scheme()
    if scheme not in ["http", "ftp", "mailto"]:
        # C:/ or something simiar:
        if os.path.splitdrive(urlPath)[0] != "":
            url = QUrl("file:///"+urlPath)
    ok = QDesktopServices.openUrl(url)
    if not ok:
        raise Exception("could not open '%s'" % url.toString())


def _recalculateMzPeakFor(postfix):
    def calculator(table, row, name, postfix=postfix):

        mzmin = table.get(row, "mzmin"+postfix)
        mzmax = table.get(row, "mzmax"+postfix)
        rtmin = table.get(row, "rtmin"+postfix)
        rtmax = table.get(row, "rtmax"+postfix)
        pm    = table.get(row, "peakmap"+postfix)
        mz = pm.representingMzPeak(mzmin, mzmax, rtmin, rtmax)
        return mz if mz is not None else (mzmin+mzmax)/2.0
    return calculator

def _hasRangeColumns(table, postfix):
    return all([table.hasColumn(n + postfix) for n in ["rtmin", "rtmax",
                                                 "mzmin", "mzmax", "peakmap"]])

def recalculateMzPeaks(table):
    #TODO: tests !
    """Adds mz value for peaks not detected with centwaves algorithm based on
       rt and mz window: needed are columns mzmin, mzmax, rtmin, rtmax and
       peakmap mz, postfixes are automaticaly taken into account"""
    postfixes = [ "" ] + [ "__%d" % i for i in range(len(table._colNames))]
    for postfix in postfixes:
        if _hasRangeColumns(table, postfix):
            mz_col = "mz" + postfix
            if table.hasColumn(mz_col):
                table.replaceColumn(mz_col, _recalculateMzPeakFor(postfix),
                                    format="%.5f", type_=float)
            else:
                table.addColumn(mz_col, _recalculateMzPeakFor(postfix),
                                format="%.5f", type_=float)

def startfile(path):
    import sys, os
    if sys.platform == "win32":
        path = path.replace("/", "\\") # needed for network paths
    return os.startfile(path)

