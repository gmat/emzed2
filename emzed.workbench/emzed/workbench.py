def run():
    print 42

def main():
    import os
    print os.system("pip install -i http://127.0.0.1:3142/root/dev/+simple emzed.core")
    import emzed.core
    print emzed.core



