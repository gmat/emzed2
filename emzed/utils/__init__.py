from utils import *
from isotope_calculator import isotopeDistributionTable, plotIsotopeDistribution
from formula_generator  import formulaTable

from load_utils import loadCSV, loadTable, loadPeakMap

from feature_detectors import runCentwave, runMatchedFilters, runMetaboFeatureFinder

from metlin import matchMetlin

from integration import integrate
