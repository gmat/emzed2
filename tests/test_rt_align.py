import os
import numpy as np

import pyopenms

IS_PYOPENMS_2 = pyopenms.__version__.startswith("2.")

import emzed

from emzed.align import rtAlign

import pytest


@pytest.fixture
def input_tables():

    def path():
        import os.path
        here = os.path.dirname(os.path.abspath(__file__))

        def j(*a):
            return os.path.join(here, *a)
        return j

    ft = emzed.io.loadTable(path()("data", "features.table"))

    ft.replaceColumn("rtmin", ft.rt - 2.0)
    ft.replaceColumn("rtmax", ft.rt + 10.0)

    # make copy and shift
    ft2 = ft.copy()

    def shift(t, col):
        ix = t.getIndex(col)
        for r in t.rows:
            r[ix] += 2.0 + 0.1 * r[ix] - 0.005 * r[ix] * r[ix]

    shift(ft2, "rt")
    shift(ft2, "rtmin")
    shift(ft2, "rtmax")

    pms = set(ft2.getValue(row, "peakmap") for row in ft2.rows)
    pmrtsbefore = []
    assert len(pms) == 1
    for pm in pms:
        for spec in pm.spectra:
            pmrtsbefore.append(spec.rt)
            spec.rt += 2.0 + 0.1 * spec.rt - 0.005 * spec.rt * spec.rt

    # delete one row, so ft should become reference map !
    del ft2.rows[-1]

    return (ft, ft2)


def test_one(input_tables, regtest, tmpdir):

    tmpdir = tmpdir.strpath

    ft, ft2 = input_tables

    ftneu, ft2neu = rtAlign([ft, ft2], refTable=ft, destination=tmpdir, nPeaks=9999,
                            numBreakpoints=3)

    check(ft, ft2, ftneu, ft2neu, tmpdir, regtest)


def test_two(input_tables, regtest, tmpdir):

    tmpdir = tmpdir.strpath
    ft, ft2 = input_tables
    ft2neu, ftneu = rtAlign([ft2, ft], refTable=ft, destination=tmpdir, nPeaks=9999,
                            numBreakpoints=3)

    check(ft, ft2, ftneu, ft2neu, tmpdir, regtest)


def test_three(input_tables, regtest, tmpdir):

    tmpdir = tmpdir.strpath
    ft, ft2 = input_tables
    ftneu, ft2neu = rtAlign([ft, ft2], destination=tmpdir, nPeaks=9999,
                            numBreakpoints=3)

    check(ft, ft2, ftneu, ft2neu, tmpdir, regtest)


def test_four(input_tables, regtest, tmpdir):
    import emzed.utils
    ft, ft2 = input_tables
    ft = emzed.utils.integrate(ft, "max")
    tmpdir = tmpdir.strpath

    # fails becaus one is integrated, after alignment the integration would be infeasible:
    with pytest.raises(Exception):
        ftneu, ft2neu = rtAlign([ft, ft2], destination=tmpdir, nPeaks=9999,
                                numBreakpoints=3)

    assert ft.method.countNotNone() > 0
    assert ft.area.countNotNone() > 0
    assert ft.params.countNotNone() > 0

    # enforces alignment, but sets integration related columns to None
    ftneu, ft2neu = rtAlign([ft, ft2], destination=tmpdir, nPeaks=9999,
                            numBreakpoints=5, resetIntegration=True)

    # all Nones:
    assert ftneu.method.countNotNone() == 0
    assert ftneu.area.countNotNone() == 0
    assert ftneu.params.countNotNone() == 0


# as we have regression tests where the output changes from pyopenms to pyopenms2
# we have to distinguish both cases. the name of the file recording the output
# of the tests depend on the functions names, so we use different function names
# in order to force recording results in different files:

if IS_PYOPENMS_2:
    pf = "_2"
else:
    pf = "_1"

test_one.__name__ = "test_one_for_pyopenms" + pf
test_two.__name__ = "test_two_for_pyopenms" + pf
test_three.__name__ = "test_three_for_pyopenms" + pf
test_four.__name__ = "test_four_for_pyopenms" + pf


def getrtsfrompeakmap(table):
    pms = set(table.getValue(row, "peakmap") for row in table.rows)
    assert len(pms) == 1
    pm = pms.pop()
    return np.array([spec.rt for spec in pm.spectra])


def check(ft, ft2, ftneu, ft2neu, path, regtest):

    alignment_should_produce_plot(path)
    realignment_should_fail(ft, ft2, ftneu, ft2neu, path, regtest)
    spectra_rt_should_be_aligned(ft, ft2, ftneu, ft2neu, path, regtest)
    rt_values_in_peak_table_should_be_aligned(ft, ft2, ftneu, ft2neu, path, regtest)


def realignment_should_fail(ft, ft2, ftneu, ft2neu, path, regtest):
    ex = None
    try:
        ftneu, ft2neu = rtAlign([ftneu, ft2neu], destination=path, nPeaks=9999,
                                numBreakpoints=2)
    except Exception, e:
        ex = e
    assert ex.message.startswith("there are already rt_aligned peakmaps")
    assert ex is not None, "aligning of aligned maps should not be possible"


def alignment_should_produce_plot(path):
    assert os.path.exists(os.path.join(path, "test_mini_aligned.png"))


def spectra_rt_should_be_aligned(ft, ft2, ftneu, ft2neu, path, regtest):
    # rts of spctra should be aligned:

    t = emzed.utils.toTable("rt_peakmap_before", getrtsfrompeakmap(ft2))
    t.addColumn("rt_peakmap_after", getrtsfrompeakmap(ft2neu))

    print >> regtest, "rt comparision of spectra in peakmaps:"
    print >> regtest, t
    print >> regtest


def rt_values_in_peak_table_should_be_aligned(ft, ft2, ftneu, ft2neu, path, regtest):

    ft = ft.extractColumns("id", "rt")
    ft2 = ft2.extractColumns("id", "rt", "rtmin", "rtmax")
    ftneu = ftneu.extractColumns("id", "rt")
    ft2neu = ft2neu.extractColumns("id", "rt", "rtmin", "rtmax")

    ft.addPostfix("_1")

    before = ft.join(ft2, ft.id_1 == ft2.id)
    before.renamePostfixes(__0="_2")
    before.dropColumns("id_2")

    compared = before.join(ftneu, before.id_1 == ftneu.id)
    compared.dropColumns("id__0")
    compared.renamePostfixes(__0="_1_aligned")

    compared = compared.join(ft2neu, compared.id_1 == ft2neu.id)
    compared.dropColumns("id__0", "id_1")
    compared.renamePostfixes(__0="_2_aligned")

    print >> regtest, "peaks before and after alignment:"
    print >> regtest, compared
