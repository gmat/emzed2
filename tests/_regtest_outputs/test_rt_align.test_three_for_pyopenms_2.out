rt comparision of spectra in peakmaps:
rt_peakmap_before rt_peakmap_after
float             float
------            ------
0.34m             0.30m
0.34m             0.30m
0.34m             0.30m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
0.34m             0.31m
... (775 rows) ...
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m
0.51m             0.50m


peaks before and after alignment:
rt_1     rt_2     rtmin_2  rtmax_2  rt_1_aligned rt_2_aligned rtmin_2_aligned rtmax_2_aligned
float    float    float    float    float        float        float           float
------   ------   ------   ------   ------       ------       ------          ------
0.49m    0.50m    0.47m    0.63m    0.49m        0.49m        0.46m           0.65m
0.31m    0.35m    0.32m    0.49m    0.31m        0.31m        0.28m           0.48m
0.50m    0.51m    0.48m    0.63m    0.50m        0.50m        0.47m           0.65m
0.47m    0.48m    0.45m    0.61m    0.47m        0.47m        0.43m           0.63m
0.50m    0.51m    0.48m    0.63m    0.50m        0.50m        0.47m           0.66m
0.46m    0.48m    0.45m    0.60m    0.46m        0.46m        0.43m           0.62m
0.47m    0.48m    0.45m    0.61m    0.47m        0.47m        0.43m           0.62m
0.45m    0.47m    0.44m    0.60m    0.45m        0.45m        0.42m           0.61m
0.33m    0.37m    0.34m    0.51m    0.33m        0.33m        0.30m           0.50m
0.32m    0.36m    0.33m    0.50m    0.32m        0.32m        0.29m           0.49m
0.35m    0.38m    0.35m    0.52m    0.35m        0.35m        0.31m           0.51m
0.46m    0.48m    0.45m    0.61m    0.46m        0.46m        0.43m           0.62m
... (3 rows) ...
0.50m    0.51m    0.48m    0.63m    0.50m        0.50m        0.46m           0.65m
0.47m    0.49m    0.46m    0.61m    0.47m        0.47m        0.44m           0.63m
0.39m    0.42m    0.39m    0.55m    0.39m        0.39m        0.36m           0.56m
0.49m    0.50m    0.47m    0.63m    0.49m        0.49m        0.46m           0.65m
0.32m    0.35m    0.32m    0.50m    0.32m        0.32m        0.28m           0.48m
0.33m    0.36m    0.33m    0.50m    0.33m        0.33m        0.29m           0.49m
0.47m    0.48m    0.45m    0.61m    0.47m        0.47m        0.43m           0.63m
0.34m    0.37m    0.34m    0.51m    0.34m        0.34m        0.30m           0.50m
0.32m    0.35m    0.32m    0.50m    0.32m        0.32m        0.28m           0.48m
0.32m    0.35m    0.32m    0.50m    0.32m        0.32m        0.28m           0.48m
0.35m    0.38m    0.35m    0.52m    0.35m        0.35m        0.32m           0.52m
0.33m    0.36m    0.33m    0.50m    0.33m        0.33m        0.29m           0.49m
0.37m    0.40m    0.37m    0.54m    0.37m        0.37m        0.34m           0.54m

